File: marsrover.exs
defmodule MarsRover do
  @moduledoc """
  For Multiverse

  This is a solution to the problem documented in the following file:
  Multiverse engineering take-home challenge (2).pdf

  Accepts input from STDIN and outputs results to STDOUT.

  Author: Iain Argent (iainargent007@gmail.com)
  """

  @doc """
  update the rover's position from its
  last location and the direction it was facing
  returns a pair of integers with the new position
  """
  def go_forward(xloc, yloc, dir) do
    case dir do
      ~c"N" -> [xloc, yloc + 1]
      ~c"E" -> [xloc + 1, yloc]
      ~c"S" -> [xloc, yloc - 1]
      ~c"W" -> [xloc - 1, yloc]
    end
  end

  @doc """
  implements a simple circular list for finding
  the next direction in a list of directions
  when rotating left or right
  returns a char with the new direction
  """
  def direction_change(dirs, curdir, isRight) do
    curpos = Enum.find_index(dirs, fn x -> x == curdir end)

    increment =
      if isRight do
        1
      else
        -1
      end

    Enum.at(dirs, rem(curpos + increment, length(dirs)))
  end

  @doc """
  recursive function which executes one instruction
  at the head of movestr
  returns a string with the report of the final location
  """
  def do_move(gridWidth, gridHeight, xloc, yloc, dir, moves) do
    dirs = [~c"N", ~c"E", ~c"S", ~c"W"]

    case moves do
      [] ->
        "(#{xloc}, #{yloc}, #{dir})"

      [curmove | tail] ->
        case curmove do
          ~c"F" ->
            [newxloc, newyloc] = go_forward(xloc, yloc, dir)
            # check if rover is out of bounds
            if newxloc < 0 or newxloc > gridWidth or newyloc < 0 or newyloc > gridHeight do
              "(#{xloc}, #{yloc}, #{dir}) LOST"
            else
              do_move(gridWidth, gridHeight, newxloc, newyloc, dir, tail)
            end

          ~c"L" ->
            do_move(gridWidth, gridHeight, xloc, yloc, direction_change(dirs, dir, false), tail)

          ~c"R" ->
            do_move(gridWidth, gridHeight, xloc, yloc, direction_change(dirs, dir, true), tail)

          _ ->
            IO.inspect(curmove)
            IO.puts("Bad input #{curmove}")
        end
    end
  end

  @doc """
  Tail recursive function which uses a regex to get
  the initial status and movement commands for a rover
  then calls do_move() to execute the movement commands.
  Finally, calls itself to check if there is more input.
  """
  def read_line(gridx, gridy) do
    case IO.gets("") do
      :eof ->
        :ok

      line ->
        regex = ~r/^\((\d+),\s*(\d+),\s*([NESW])\)\s*([LRF]*)$/

        case Regex.run(regex, line) do
          nil ->
            IO.puts("Invalid input. Please enter a command like '(2, 3, E) LFRFF'.")

          [_, xloc, yloc, dir, movestr] ->
            # IO.puts("You entered: #{xloc}, #{yloc} facing #{dir} with #{movestr}")
            result =
              MarsRover.do_move(
                gridx,
                gridy,
                String.to_integer(xloc),
                String.to_integer(yloc),
                String.to_charlist(dir),
                String.to_charlist(movestr) |> Enum.map(&[&1])
              )

            IO.puts(result)
        end

        read_line(gridx, gridy)
    end
  end

  @doc """
  Get the grid size from the first line,
  then call read_line() to process other input
  """
  def main() do
    gridstr = IO.gets("")
    numbers = String.trim(gridstr) |> String.split()
    [gridx, gridy] = Enum.map(numbers, &String.to_integer/1)
    # IO.puts("You entered: #{gridx} and #{gridy}")
    read_line(gridx, gridy)
  end
end

MarsRover.main()

File: marsrover.py
"""
For Multiverse

This is a solution to the problem documented in the following file:
Multiverse engineering take-home challenge (2).pdf

Author: Iain Argent (iainargent007@gmail.com)

Approach:
This was written in a functional style (as opposed to object-oriented)
using a recursive function to process one movement at a time.
Although Python does not support tail recursion other languages do
and an equivalent implementation in Elixir would be tail recursive.
I chose Python to complete this challenge as it is fairly easy to read
and I know it better than Elixir. 

Usage:
You can use this program in interactive mode like this:

Windows:
python.exe marsrover.py

Mac/BSD/Linux:
python3 marsrover.py

Or you can save commands in a file (e.g. marsrove.txt)
and execute all the commands like this:

Windows:
type .\marsrover.txt | python.exe marsrover.py

Mac/BSD/Linux:
cat marsrover.txt | python3 marsrover.py

Test suite:
There is an accompanying test suite called test_marsrover.py
"""

import re


# implements a simple circular list for finding
# the next direction in a list of directions
# when rotating left or right
# returns a char with the new direction
def direction_change(dirs: list, curdir: str, isRight: bool):
    curpos = dirs.index(curdir)
    increment = 1 if isRight else -1
    return dirs[(curpos + increment) % len(dirs)]


# update the rover's position from its
# last location and the direction it was facing
# returns a pair of integers with the new position
def go_forward(xloc: int, yloc: int, dir: str):
    if dir == "N":
        return (xloc, yloc + 1)
    if dir == "E":
        return (xloc + 1, yloc)
    if dir == "S":
        return (xloc, yloc - 1)
    if dir == "W":
        return (xloc - 1, yloc)
    raise ValueError("Invalid direction: " + dir)


# recursive function which executes one instruction
# at the head of movestr
# returns a string with the report of the final location
def do_move(
    xloc: int, yloc: int, dir: str, gridWidth: int, gridHeight: int, movestr: list
):
    dirs = ["N", "E", "S", "W"]
    if movestr == []:
        return f"({xloc}, {yloc}, {dir})"
    else:
        curmove, *tailstr = movestr
        if curmove not in ["L", "R", "F"]:
            raise ValueError("Invalid command found: " + curmove)
        if curmove == "F":
            (newxloc, newyloc) = go_forward(xloc, yloc, dir)

            # check if rover is out of bounds
            if (
                newxloc < 0
                or newxloc > (gridHeight - 1)
                or newyloc < 0
                or newyloc > (gridHeight - 1)
            ):
                return f"({xloc}, {yloc}, {dir}) LOST"
            else:
                return do_move(newxloc, newyloc, dir, gridWidth, gridHeight, tailstr)
        else:
            newdir = direction_change(dirs, dir, curmove == "R")
        return do_move(xloc, yloc, newdir, gridWidth, gridHeight, tailstr)


if __name__ == "__main__":
    gridstr = input()

    # retrieve the two numbers from the string (as integers)
    (gridx, gridy) = map(int, gridstr.split())

    while True:
        try:
            comstr = input()
        except EOFError:
            break
        pattern = r"\(([^,]+)\W*([^,]+).*\W*([^\)]+)\)\W([LRF]*)"
        (xloc, yloc, dir, movestr) = re.findall(pattern, comstr)[0]
        result = do_move(int(xloc), int(yloc), dir, gridx, gridy, list(movestr))
        print(result)
